dex/src/main/scala/com/wavesplatform/dex/SpendableBalancesActor.scala

```diff
@@ -82,20 +85,40 @@ class SpendableBalancesActor(spendableBalances: (Address, Set[Asset]) => Future[
     case SpendableBalancesActor.Command.UpdateStates(changes) =>
       changes.foreach {
         case (address, stateUpdate) =>
+          val updateAndSendChanges = updateStateAndSendChanges(address, stateUpdate) _
           fullState.get(address) match {
-            case Some(addressFullState) =>
-              val cleanStateUpdate = stateUpdate.filter { case (a, updatedBalance) => addressFullState.get(a).fold(true)(updatedBalance != _) }
-              addressDirectory ! AddressDirectory.Envelope(address, AddressActor.Command.CancelNotEnoughCoinsOrders(cleanStateUpdate))
-              fullState = fullState.updated(address, addressFullState ++ cleanStateUpdate)
+            case Some(afs) => fullState = updateAndSendChanges(fullState, afs)(true)
             case None =>
-              addressDirectory ! AddressDirectory.Envelope(address, AddressActor.Command.CancelNotEnoughCoinsOrders(stateUpdate))
-              incompleteStateChanges = incompleteStateChanges.updated(address, incompleteStateChanges.getOrElse(address, Map.empty) ++ stateUpdate)
+              incompleteStateChanges.get(address) match {
+                case Some(ais) => incompleteStateChanges = updateAndSendChanges(incompleteStateChanges, ais)(true)
+                case None      => incompleteStateChanges = updateAndSendChanges(incompleteStateChanges, Map.empty)(false)
+              }
```

Comment:
It looks compact but it is hard to read (I am probably just tired). IMO the main reason is `updateStateAndSendChanges`:
1. works both with the state and sends the message
2. is over-abstracted. It has 5 arguments and only 4 lines of logic

I tried to rework this code (also see comments below) and what I got:
```scala
    case SpendableBalancesActor.Command.UpdateStates(balanceChanges) =>
      balanceChanges.foreach {
        case (address, balanceChanges) =>
          val origFullState               = fullState.get(address)
          lazy val origIncompleteBalances = incompleteStateChanges.get(address)

          // Review comment: here we change the state
          origFullState match {
            case origBalance @ Some(origBalances) => fullState = fullState.updated(address, origBalances ++ balanceChanges)
            case None =>
              val updatedAddressState = origIncompleteBalances.foldLeft(balanceChanges)(_ ++ _)
              incompleteStateChanges = incompleteStateChanges.updated(address, updatedAddressState)
          }

          // Review comment: here we send the message to AA
          val knownBalance = origFullState.orElse(origIncompleteBalances).getOrElse(Map.empty)
          addressDirectory ! mkBalanceChanged(knownBalance, balanceChanges)
      }

  // ...

  // Review comment: we have duplicated BalanceChanged.apply here, but mkBalanceChanged looks easier than getCleanBalancesAndAssetsToCheck
  private def mkBalanceChanged(knownBalance: AddressState, balanceChanges: AddressState): BalanceChanged =
    if (knownBalance.isEmpty) BalanceChanged(balanceChanges, balanceChanges)
    else {
      val (cleanBalanceChanges, balancesToCheck) = balanceChanges.foldLeft { (Map.empty: AddressState, Map.empty: AddressState) } {
        case ((cleanBalanceChanges, balancesToCheck), balanceChange @ (asset, updatedBalance)) =>
          knownBalance.get(asset) match {
            case None => (cleanBalanceChanges + balanceChange, balancesToCheck + balanceChange)
            case Some(knownBalance) =>
              if (knownBalance == updatedBalance) (cleanBalanceChanges, balancesToCheck)
              else {
                val updatedAssetsToCheck = if (updatedBalance < knownBalance) balancesToCheck + balanceChange else balancesToCheck
                (cleanBalanceChanges + balanceChange, updatedAssetsToCheck)
              }
          }
      }
      BalanceChanged(cleanBalanceChanges, balancesToCheck)
    }
```